bolt-modules/boltlib/lib/puppet/functions/puppetdb_fact.rb

```diff
@@ -0,0 +1,27 @@
+# frozen_string_literal: true
+
+require 'bolt/error'
+
+# Returns a hash of certname to facts objects for each matched Target.
+# This functions accepts a TargetSpec, and returns a hash of target certnames
+# and their corresponding facts hash.
+#
+# * If a node is not found in PuppetDB, it's included in the returned hash with empty facts hash.
+# * Otherwise the node is included in the hash with a value that is a hash of it's facts.
+#
+Puppet::Functions.create_function(:puppetdb_fact) do
+  dispatch :puppetdb_fact do
+    param 'Boltlib::TargetSpec', :targets
+  end
+
+  def puppetdb_fact(targets)
+    unless Puppet[:tasks]
+      raise Puppet::ParseErrorWithIssue.from_issue_and_stack(
+        Puppet::Pops::Issues::TASK_OPERATION_NOT_SUPPORTED_WHEN_COMPILING, operation: 'puppetdb_fact'
+      )
+    end
+
+    executor = Puppet.lookup(:bolt_executor) { nil }
+    executor.puppetdb_fact(targets)
```

Comment:
I don't really like the idea of everything being hard-coded into the executor. However it does make sense for some things to be attached to its lifetime.

I'd probably be in favor of a caching mechanism to store and retrieve function-specific data in the executor. I'd probably enforce a namespace (the function name) and key for lookup.

Reply:
I guess in this case we also need access to the global config. Does it make sense to give all functions access to that?

Reply:
I guess this is a generalization problem we can address later when it becomes necessary.

Reply:
I actually initially had this live in the puppetdb client class, but the issue is that then every time `run_plan` is called it inits a connection to puppetdb whether we need it or not. I could refactor the pdb client to just init a 'dummy' connection, then actually initialize it if a function that needs it is called, but this seemed easier + more like "the right way". But would that actually be preferable?

I'm not sure I understand what 'namespace' means in that context? Would the caching just be a hash, and what data would it store? Or did you have something else in mind?

Reply:
I meant something like
```
dbclient = executor.cache_lookup(:puppetdb_fact, :client)
unless dbclient
  dbconfig = Bolt::PuppetDB::Config.new(executor.config.puppetdb)
  dbclient = Bolt::PuppetDB::Client.from_config(dbconfig)
  executor.cache_store(:puppetdb_fact, :client, dbclient)
end

dbclient.facts_for_node(certnames)
```

But since we're not exposing config, I don't think we should do that until we actually need it. Hard-coding puppetdb as a thing the executor handles is fine for now.

Reply:
Ah, ok, I see what you mean. M'k, that seems reasonable.

Reply:
where "that" is waiting until...we need to limit access to parts of the configuration?

Reply:
I don't think we should change what you have and add support for generalized caching on the executor until we have better use cases. I don't want to open up direct access to config from Bolt functions without a better reason.